File: src/com/interview/dynamic/EditDistance.java
package com.interview.dynamic;

/**
 http://www.geeksforgeeks.org/dynamic-programming-set-5-edit-distance/
 */
public class EditDistance {

    private static int EDIT_COST = 1;
    
    public int recEditDistance(char[]  str1, char str2[], int len1,int len2){
        
        if(len1 == str1.length){
            return str2.length - len2;
        }
        if(len2 == str2.length){
            return str1.length - len1;
        }
        return min(recEditDistance(str1,str2,len1+1,len2+1) + str1[len1] == str2[len2] ? 0 : 1, recEditDistance(str1,str2,len1,len2+1) + 1, recEditDistance(str1,str2,len1+1,len2) + 1);
    }
    
    private int min(int a,int b, int c){
        int l = Math.min(a, b);
        return Math.min(l, c);
    }
    
    
    public int dynamicEditDistance(char[] str1, char[] str2){
        int temp[][] = new int[str1.length+1][str2.length+1];
        
        for(int i=0; i < temp[0].length; i++){
            temp[0][i] = i;
        }
        
        for(int i=0; i < temp.length; i++){
            temp[i][0] = i;
        }
        
        for(int i=1;i <=str1.length; i++){
            for(int j=1; j <= str2.length; j++){
                if(str1[i-1] == str2[j-1]){
                    temp[i][j] = temp[i-1][j-1];
                }else{
                    temp[i][j] = min(temp[i-1][j-1] + 1,temp[i-1][j] + EDIT_COST,temp[i][j-1] + EDIT_COST);
                }
            }
        }
        return temp[str1.length][str2.length];
        
    }
    
    public static void main(String args[]){
        String str1 = "SUNDAY";
        String str2 = "SATURDAY";
        EditDistance editDistance = new EditDistance();
        int result = editDistance.dynamicEditDistance(str1.toCharArray(), str2.toCharArray());
        System.out.print(result);
    }

}

File: python/dynamic/editdistance.py
#edit distance dynamic programming
# java code https://github.com/mission-peace/interview/blob/master/src/com/interview/dynamic/EditDistance.java
def bottom_up_dp(str1, str2):
    m = len(str1) + 1
    n = len(str2) + 1
    T = [[0 for x in range(n)] for x in range(m)]
    for i in range(n):
        T[0][i] = i
        
    for i in range(m):
        T[i][0] = i

    for i in range(1, m):
        for j in range(1, n):
            if str1[i-1] == str2[j-1]:
                T[i][j] = T[i-1][j-1]
            else:
                T[i][j] = 1 + min(T[i-1][j-1], T[i-1][j], T[i][j-1])

    print_actual_edits(T, str1, str2)            
    return T[m-1][n-1]

def print_actual_edits(T, str1, str2):
    i = len(T) - 1
    j = len(T[0]) - 1
    while True:
        if i == 0 or j == 0:
            break
        if str1[i-1] == str2[j-1]:
            i = i-1
            j = j-1
        elif T[i][j] == T[i-1][j-1] + 1:
            print("Edit " + str2[j-1] + " in string2 to " + str1[i-1] + " in string1")
            i = i - 1
            j = j - 1
        elif T[i][j] == T[i-1][j] + 1:
            print("Delete in string1 " + str1[i-1])
            i = i - 1
        elif T[i][j] == T[i][j-1] + 1:
            print("Delete in string2 " + str2[j-1])
            j = j - 1
    
if __name__ == '__main__':
    result = bottom_up_dp(list('azced'), list('abcdef'))
    print(result)
